RF: Use python properties, rather than set/get methods

bcipolli opened the pull request:
`numDA`, `version`, and `filename` properties on GIFTI have `set_` and `get_` methods. Seems cleaner (and more pythonic) to use Python properties instead.

Question: is GIFTI support officially published? I.e. do I need to add deprecation warnings for the old methods, and deprecation tests?

Also renamed tests to be clear that they're testing the loading of these properties, not necessarily of their setting or saving.

matthew-brett commented:
Sorry to say, but yes, we have advertised the GIFTI support, and people have used it, so I think we do have to deprecate the get_ / set_ methods...

bcipolli commented:
OK. Is it worth it to make this change, in this case? Happy to deprecate & test, but also OK to toss. I'm just working through the code on my way to load/save GIFTI.

matthew-brett commented:
If you have the heart to do the deprecations, I think it's an improvement...

bcipolli commented:
I sure do! Will get to that tonight or tomorrow, then :)

bcipolli commented:
Added the get/set deprecations and deprecation tests.

matthew-brett commented on nibabel/gifti/gifti.py:
```diff
@@ -427,14 +425,12 @@ def add_gifti_data_array(self, dataarr):
         """
         if isinstance(dataarr, GiftiDataArray):
             self.darrays.append(dataarr)
-            self.numDA += 1
```
numDA being tested properly already?

bcipolli replied:
Not sure about tests, will check.
This line was removed as I made `numDA` a computed read-only property.

matthew-brett replied:
Yes, sorry to be confusing - I saw you didn't need the line, but I was wondering if there were tests to make sure the computation was working right.

bcipolli replied:
There were not; I just added some.

bcipolli commented:
I updated some other functions that are better as properties and added basic tests. Please take another look when possible!

bcipolli commented on nibabel/gifti/gifti.py:
```diff
@@ -392,15 +403,24 @@ def set_labeltable(self, labeltable):
-        else:
-            print("Not a valid GiftiLabelTable instance")
+        if not isinstance(labeltable, GiftiLabelTable):
+            raise ValueError("Not a valid GiftiLabelTable instance")
```
Note the change from `print` to `ValueError` here.

effigies commented on nibabel/gifti/gifti.py:
```diff
@@ -128,10 +133,31 @@ def __init__(self, key=0, label='', red=None, green=None, blue=None,
+        self.red = rgba[0]
+        self.green = rgba[1]
+        self.blue = rgba[2]
+        self.alpha = rgba[3]
```
Personal preference, feel free to reject:

    self.red, self.green, self.blue, self.alpha = rgba

bcipolli replied:
Good call!

effigies commented on nibabel/gifti/parse_gifti_fast.py:
```diff
@@ -108,9 +108,6 @@ def StartElementHandler(self, name, attrs):
                 self.img.version = attrs['Version']
-            if 'NumberOfDataArrays' in attrs:
-                self.img.numDA = int(attrs['NumberOfDataArrays'])
-                self.count_da = False
```
I think `attrs['NumberOfDataArrays']` should at least be validated and throw a warning if the count doesn't match the actual number of data arrays. Something like:

    if 'NumberOfDataArrays' in attrs:
        self.expected_numda = int(attrs['NumberOfDataArrays'])

bcipolli replied:
Yes. I will also add a test for this. In fact, we lack malformed GIFTI file tests; see https://github.com/nipy/nibabel/issues/357

effigies commented on nibabel/gifti/tests/test_gifti.py:
```diff
@@ -24,6 +26,35 @@ def test_gifti_image():
+    gi.darrays[0].intent = 0
+    gi.remove_gifti_data_array_by_intent(0)
+    assert_equal(gi.numDA, 0)
+
```
Maybe a check that trying to remove a data array when `numDA == 0` produces an `IndexError` and maintains a `numDA == 0`? At least with the old version where the number was incremented and decremented manually, I'd want to be sure it can't go negative.

bcipolli replied:
I believe that test is here: https://github.com/nipy/nibabel/pull/353#discussion-diff-42019172R42

effigies replied:
Wow. Yup. Now I'm questioning my ability to read. Should probably have another pass through in a bit.

effigies commented on nibabel/gifti/tests/test_gifti.py:
```diff
@@ -36,3 +67,58 @@ def test_dataarray():
+    assert_equal(kwargs['red'], gl.rgba[0])
+    assert_equal(kwargs['green'], gl.rgba[1])
+    assert_equal(kwargs['blue'], gl.rgba[2])
+    assert_equal(kwargs['alpha'], gl.rgba[3])
```
Why not just:

     assert_equal(rgba, gl.rgba)

bcipolli replied:
Doesn't work for me; I get:
```ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()```

But `numpy.assert_all_equal` works :)

effigies commented:
Read-through done. Logic looks good.

bcipolli commented:
Thanks @effigies ; I just updated the `rgba` logic and tests, added requested check for `NumberOfDataArrays` (and added a test), and opened #357 so we can add more 'malformed' GIFTI tests.

bcipolli commented:
Updated based on last comments.

effigies commented on nibabel/gifti/tests/test_gifti.py:
```diff
@@ -36,3 +67,63 @@ def test_dataarray():
+
+    # Test default value
+    gl4 = GiftiLabel()
+    assert_is_none(gl4.rgba[0])
```
2.6 seems upset.

    assert_true(gl4.rgba[0] is None)

bcipolli replied:
That's what I get for trying to be too fancy :)

effigies commented:
Okay, this LGTM.

@matthew-brett The Travis build seems to be failing on unrelated things. (Happening here and on #354.)
